code/datums/components/pellet_cloud.dm

```diff
@@ -0,0 +1,249 @@
+		else if(radius >= 2)
+			L.visible_message("<b><span class='danger'>[L] heroically covers \the [parent] with [L.p_their()] body, absorbing some of the shrapnel!</span></b>", "<span class='userdanger'>You heroically cover \the [parent] with your body, absorbing some of the shrapnel!</span>")
+			magnitude_absorbed += 2
+		else
+			L.visible_message("<b><span class='danger'>[L] heroically covers \the [parent] with [L.p_their()] body, snuffing out the shrapnel!</span></b>", "<span class='userdanger'>You heroically cover \the [parent] with your body, snuffing out the shrapnel!</span>")
+			magnitude_absorbed = radius
+
+		var/pellets_absorbed
+		pellets_absorbed = (radius ** 2) - ((radius - magnitude_absorbed - 1) ** 2)
+		radius -= magnitude_absorbed
+		total_pellets_absorbed += round(pellets_absorbed/2)
+
+		if(L.stat != DEAD && L.client)
+			LAZYADD(purple_hearts, L)
+
+		for(var/i in 1 to round(pellets_absorbed/2))
+			pew(L)
+
+		if(radius < 1)
+			break
+
+	return total_pellets_absorbed
+
+///One of our pellets hit something, record what it was and check if we're done (terminated == num_pellets)
+/datum/component/pellet_cloud/proc/pellet_hit(obj/projectile/P, atom/movable/firer, atom/target, Angle)
+	pellets -= P
+	terminated++
+	hits++
+	targets_hit[target]++
+	UnregisterSignal(P, list(COMSIG_PARENT_QDELETING, COMSIG_PROJECTILE_RANGE_OUT, COMSIG_PROJECTILE_SELF_ON_HIT))
+	if(terminated == num_pellets)
+		finalize()
+
+///One of our pellets disappeared due to hitting their max range (or just somehow got qdel'd), remove it from our list and check if we're done (terminated == num_pellets)
+/datum/component/pellet_cloud/proc/pellet_range(obj/projectile/P)
+	pellets -= P
+	terminated++
+	UnregisterSignal(P, list(COMSIG_PARENT_QDELETING, COMSIG_PROJECTILE_RANGE_OUT, COMSIG_PROJECTILE_SELF_ON_HIT))
+	if(terminated == num_pellets)
+		finalize()
+
+/// Minor convenience function for creating each shrapnel piece with circle explosions, mostly stolen from the MIRV component
+/datum/component/pellet_cloud/proc/pew(atom/target, spread=0)
+	var/obj/projectile/P = new projectile_type(get_turf(parent))
+
+	//Shooting Code:
+	P.spread = spread
+	P.original = target
+	P.fired_from = parent
+	P.firer = parent // don't hit ourself that would be really annoying
+	P.permutated += parent // don't hit the target we hit already with the flak
+	P.suppressed = SUPPRESSED_VERY // set the projectiles to make no message so we can do our own aggregate message
+	P.preparePixelProjectile(target, parent)
+	RegisterSignal(P, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
+	RegisterSignal(P, list(COMSIG_PROJECTILE_RANGE_OUT, COMSIG_PARENT_QDELETING), .proc/pellet_range)
+	pellets += P
+	P.fire()
+
+///All of our pellets are accounted for, time to go target by target and tell them how many things they got hit by.
+/datum/component/pellet_cloud/proc/finalize()
+	var/obj/projectile/P = new projectile_type
+	var/proj_name = initial(P.name)
+	QDEL_NULL(P)
```

Comment:
No need to create and delete a projectile to get its name. You can do it like this:
```cpp
var/obj/projectile/proj
var/proj_name = initial(proj.name)
```